File: client/src/types/product.ts
/**
 * @file product.ts
 * @description 定义商品相关的数据类型，包括商品信息、商品状态、分类、成色等常量映射
 * @input 无
 * @output Product, UnavailableProduct, CreateProductDto 等类型及商品常量
 */

/** 商品审核状态类型：待审核、已通过、已拒绝、已禁用 */
export type ProductStatus = 'PENDING' | 'APPROVED' | 'REJECTED' | 'DISABLED'

/**
 * 商品数据结构，表示一个完整的商品信息
 */
export interface Product {
  /** 商品 ID */
  id: string
  /** 商品名称 */
  name: string
  /** 商品描述 */
  description: string
  /** 商品图片 URL 列表 */
  imageUrls: string[]
  /** 商品价格 */
  price: number
  /** 库存数量 */
  quantity: number
  /** 商品分类 */
  category: string
  /** 商品成色/新旧程度 */
  condition: string
  /** 商品来源 */
  source: string
  /** 商品审核状态 */
  status: ProductStatus
  /** 拒绝原因，仅在状态为 REJECTED 时有值 */
  rejectReason: string | null
  /** 卖家信息，可能为空 */
  seller: {
    /** 卖家用户 ID */
    id: string
    /** 卖家用户名 */
    username: string
  } | null
}

/**
 * 不可用商品数据结构，表示库存不足等原因无法购买的商品
 */
export interface UnavailableProduct {
  /** 不可用原因说明 */
  message: string
  /** 商品 ID */
  productId: string
  /** 用户请求购买的数量 */
  requestedAmount: number
  /** 实际可用库存数量 */
  availableAmount: number
}

/**
 * 创建商品的数据传输对象
 */
export interface CreateProductDto {
  /** 商品名称 */
  name: string
  /** 商品描述 */
  description: string
  /** 商品图片 URL 列表 */
  imageUrls: string[]
  /** 商品价格 */
  price: number
  /** 库存数量 */
  quantity: number
  /** 商品分类 */
  category: string
  /** 商品成色 */
  condition: string
  /** 商品来源 */
  source: string
}

/** 商品分类映射：英文键 -> 中文显示名 */
export const PRODUCT_CATEGORIES: Record<string, string> = {
  ANIME_FIGURE: '手办',
  POSTER: '海报',
  KEYCHAIN: '钥匙扣',
  BADGE: '徽章',
  PILLOW: '抱枕',
  STAND: '立牌',
  CLOTHING: '服饰',
  ALBUM: '专辑',
  ACCESSORY: '配件',
  OTHER: '其他'
}

/** 商品成色映射：英文键 -> 中文显示名 */
export const PRODUCT_CONDITIONS: Record<string, string> = {
  NEW: '全新',
  LIKE_NEW: '几乎全新',
  GOOD: '良好',
  FAIR: '一般'
}

/** 商品审核状态映射：英文键 -> 中文显示名 */
export const PRODUCT_STATUSES: Record<string, string> = {
  PENDING: '待审核',
  APPROVED: '已通过',
  REJECTED: '已拒绝',
  DISABLED: '已禁用'
}

/** 商品审核状态对应的 Tailwind CSS 样式类映射 */
export const PRODUCT_STATUS_COLORS: Record<string, string> = {
  PENDING: 'bg-yellow-100 text-yellow-800',
  APPROVED: 'bg-green-100 text-green-800',
  REJECTED: 'bg-red-100 text-red-800',
  DISABLED: 'bg-gray-100 text-gray-800'
}

File: server/src/main/java/me/code/springboot_postgres/models/entities/Product.java
/**
 * @file Product.java
 * @description 商品实体类，表示平台上的商品信息
 * @input 商品名称、描述、图片、价格、数量、分类、成色、来源
 * @output 持久化的商品记录
 */
package me.code.springboot_postgres.models.entities;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UuidGenerator;
import org.hibernate.type.SqlTypes;

import java.math.BigDecimal;
import java.util.List;

/**
 * 商品实体
 * 职责：映射商品数据，包含商品基本信息、状态、审核信息和乐观锁版本控制
 */
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "products",
        indexes = {
            @Index(name = "idx_products_seller_id", columnList = "seller_id"),
            @Index(name = "idx_products_status", columnList = "status")
        })
public class Product {

    // 商品唯一标识
    @Id
    @GeneratedValue
    @UuidGenerator
    private String id;

    // 商品名称
    @Column(nullable = false)
    private String name;

    // 商品描述
    @Column(columnDefinition = "TEXT")
    private String description;

    // 商品图片URL列表（JSON格式存储）
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private List<String> imageUrls;

    // 商品价格
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    // 库存数量
    @Column(nullable = false)
    private int quantity;

    // 商品分类
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private Category category;

    // 商品成色
    @Enumerated(EnumType.STRING)
    @Column(length = 15)
    private Condition condition;

    // 商品来源（PLATFORM/USER）
    @Column(length = 20)
    private String source;

    // 商品审核状态，默认为已审核
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private Status status = Status.APPROVED;

    // 驳回原因
    @Column(name = "reject_reason", columnDefinition = "TEXT")
    private String rejectReason;

    // 商品卖家（懒加载）
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "seller_id")
    private User seller;

    // 乐观锁版本号，防止并发修改冲突
    @Version
    private int version;

    /**
     * 构造商品对象
     * @param name 商品名称
     * @param description 商品描述
     * @param imageUrls 图片URL列表
     * @param price 价格
     * @param quantity 库存数量
     * @param category 分类
     * @param condition 成色
     * @param source 来源
     */
    public Product(String name, String description, List<String> imageUrls, BigDecimal price, int quantity, Category category, Condition condition, String source) {
        this.name = name;
        this.description = description;
        this.imageUrls = imageUrls;
        this.price = price;
        this.quantity = quantity;
        this.category = category;
        this.condition = condition;
        this.source = source;
    }

    /**
     * 商品成色枚举
     */
    public enum Condition {
        NEW, LIKE_NEW, GOOD, FAIR
    }

    /**
     * 商品分类枚举
     */
    public enum Category {
        ANIME_FIGURE, POSTER, KEYCHAIN, BADGE, PILLOW, STAND, CLOTHING, ALBUM, ACCESSORY, OTHER
    }

    /**
     * 商品审核状态枚举
     */
    public enum Status {
        PENDING, APPROVED, REJECTED, DISABLED
    }
}
